### PR TITLE
feat(scheduler): once trigger + lifecycle notifications (#415)

### LINKED ISSUE #415: Support one-shot scheduled tasks + complete task lifecycle notifications
## Goal

把定时任务补到完整生命周期：不仅支持一次性执行，还要让关键状态变迁都有统一系统反馈。

## Implemented In

- PR #416（pending merge）

## Phase 1: once trigger

- [x] `{ type: 'once', fireAt: number }` 一次性延迟执行
- [x] 执行后自动退役（删库 + runtime 注销）
- [x] 重启后过期任务取消 + 通知（`SKIP_MISSED_WINDOW`）
- [x] 仅在真实执行后退役，不因治理 skip 误退役
- [x] once 任务在治理 skip 后可正确重试
- [x] 长延迟 once 任务通过 chunked timer 避免 Node `setTimeout` 溢出

## Phase 2: 生命周期通知

- [x] 注册成功 → 通知用户 + 下次执行时间
- [x] 暂停 → 通知用户
- [x] 恢复 → 通知用户 + 下次执行时间
- [x] 删除 → 通知用户
- [x] 执行成功（recurring）→ 通知用户本次完成 + 下次执行时间
- [x] 执行成功（once）→ 通知用户任务已自动结束
- [x] 执行失败 → 主动推送通知
- [x] 重启错过执行窗口（once）→ 通知用户任务已结束

## Review-Driven Hardening Included In Scope

- [x] `delayMs` / `fireAt` 做有限数校验
- [x] resume 失败时回滚持久化 `enabled=true`
- [x] runtime unregistration 后对 once timer callback 加护栏

## Why

一次性任务和周期任务本质上是同一套调度生命周期，只是执行频次不同。要补的不是单独一个 trigger，而是完整的生命周期闭环：创建、等待、执行、反馈、恢复、结束。

Related: F139 Unified Schedule Abstraction

### PR DESCRIPTION
**Pull request by mindfn**
## What

- Add `{ type: 'once', fireAt: number }` as a first-class trigger for one-shot scheduled tasks
- Complete lifecycle notifications for scheduled tasks:
  - register
  - pause
  - resume
  - delete
  - execution success
  - execution failure
  - missed execution window after restart
- Keep one-shot task semantics correct across restart and governance paths:
  - auto-retire only after an actual execution
  - cancel overdue hydrated `once` tasks instead of replaying them
  - retry `once` tasks correctly after governance skips
- Harden runtime behavior found during review:
  - cap long `setTimeout` waits via chunked re-scheduling
  - validate `delayMs` / `fireAt` as finite numbers
  - roll back persisted `enabled=true` when resume fails
  - guard once-task callbacks after runtime unregistration

## Why

The original scheduler only handled recurring triggers and left lifecycle feedback incomplete. Users could not create a true one-shot task, and they had no consistent system notification for key state transitions. This PR closes that gap so scheduled tasks behave like a full lifecycle feature instead of a partial mechanism.

## Issue Closure

- Closes #415

## Original Requirements

- **原始需求摘录**：
  > "当前的定时任务好像是只有周期执行；没有那种一次性的定时任务的；比如两分钟后帮我查询下天气并通知我"
  >
  > "设置或者执行完成后是不是系统自动触发一个通知一下用户下次的执行时间的；如果因为重启导致定时任务错过了应该自动取消并通知用户"
- 铲屎官核心痛点：
  - 定时任务不支持一次性执行
  - 生命周期状态变更缺少统一系统反馈

## Tradeoffs

- Persist `once` as absolute `fireAt` epoch ms instead of relative delay to avoid restart drift
- Remove completed one-shot tasks instead of tombstoning because `run_ledger` is already the audit trail
- Deliver lifecycle notifications fire-and-forget so notification failure does not block the primary task flow

## Scope Covered

- `packages/api/src/infrastructure/scheduler/types.ts`
- `packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts`
- `packages/api/src/infrastructure/scheduler/execute-pipeline.ts`
- `packages/api/src/infrastructure/scheduler/schedule-notify.ts`
- `packages/api/src/routes/schedule.ts`
- `packages/api/src/index.ts`
- `packages/mcp-server/src/tools/schedule-tools.ts`
- `packages/web/src/components/workspace/schedule-helpers.ts`
- `cat-cafe-skills/schedule-tasks/SKILL.md`
- scheduler test coverage for once-trigger, lifecycle notifications, missed-window behavior, and review-driven edge cases

## Verification

- `pnpm --filter @cat-cafe/api test`
- `pnpm -r --if-present run build`
- `pnpm check`
- `pnpm lint`
- PR CI checks green on latest head

## Notes

- Non-blocking test debt remains for the resume-template-missing branch in `schedule.ts`; current behavior is fixed, but there is not yet a dedicated regression test for that route branch.
- Cloud review findings from intermediate rounds were addressed in-PR; the final review returned no major issues.

---

**本地 Review**: [x] gpt52 已 review 并放行（多轮）
**云端 Review**: [x] latest round 无 major issues

### CODE REVIEW

**Comment by mindfn** [comment]
@codex review

Please review latest commit 7c68cc41 for P1/P2 only.

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

File: packages/api/src/infrastructure/scheduler/schedule-notify.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  return ((def.params as Record<string, unknown>).triggerUserId as string) ?? 'system';
+}
+
+function label(def: DynamicTaskDef): string {
+  return def.display?.label ?? def.templateId;
+}
+
+function fire(deliver: DeliverFn | undefined, def: DynamicTaskDef, content: string): void {
+  if (!deliver || !def.deliveryThreadId) return;
+  deliver({ threadId: def.deliveryThreadId, content, catId: def.createdBy, userId: resolveUserId(def) }).catch(
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/schedule-notify.ts]: Valid observation. `createdBy` is client-provided in the existing `POST /api/schedule/tasks` endpoint (line 328) — this is a pre-existing trust-the-client pattern, not introduced by the notification code. The notification just surfaces what is already stored. Fixing this would mean changing the task creation contract, which is out of scope for this PR. Acknowledged as a follow-up item.

**Comment by mindfn** [comment]
@codex review

Please review latest commit cb09a534 for P1/P2 only.

This commit addresses the 2 P1s and 1 P2 from the previous round:
- P1 fix: setTimeout delay capped at 2^31-1 ms with chunked re-scheduling
- P1 fix: setEnabled rolled back on failed resume (template missing)
- P2 fix: fireAt/delayMs validated as finite positive numbers
- P2 (createdBy): acknowledged as pre-existing pattern, out of scope (see inline reply)

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

File: packages/api/src/infrastructure/scheduler/schedule-notify.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import type { DynamicTaskDef } from './DynamicTaskStore.js';
+import type { DeliverOpts, TriggerSpec } from './types.js';
+
+type DeliverFn = (opts: DeliverOpts) => Promise<string>;
+
+/** Compute epoch ms of next fire time for a trigger */
+export function computeNextFireTime(trigger: TriggerSpec): number | null {
+  if (trigger.type === 'once') return trigger.fireAt;
+  if (trigger.type === 'cron') return Date.now() + getNextCronMs(trigger.expression, trigger.timezone);
+  if (trigger.type === 'interval') return Date.now() + trigger.ms;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/schedule-notify.ts:18]: Acknowledged as P3, not P2. This is a cosmetic approximation in notification text, not a logic error:

1. `Date.now() + ms` is the same formula already used in `notifyTaskRegistered` and `notifyTaskResumed` (both shipped without being flagged).
2. Node `setInterval` does not expose its internal anchor timestamp, so computing the exact next tick would require tracking registration time per task — added complexity for marginal notification accuracy.
3. The notification is informational only — it does not affect scheduling behavior.
4. For typical task durations (sub-second gate+execute), the error is negligible.

Will not fix in this PR. If precision becomes user-facing important, we can track `lastTickAt` in a future iteration.

**Comment by mindfn** [comment]
@codex review

Please review latest commit 1cf3c583 for P1/P2 only.

Fixes from previous round:
- P1: Once tasks no longer retire on governance skip (SKIP_GLOBAL_PAUSE/SKIP_TASK_OVERRIDE). Uses onceTaskExecuted tracking set — only retires after actual work item execution.
- P2: retireOnceTask now uses taskId directly instead of dynamicTaskIds map lookup, preventing orphaned DB rows when task is paused during in-flight execution.

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

**Comment by mindfn** [comment]
@codex review

Please review latest commit ecb1eba0 for P1/P2 only.

Fixes from previous round:
- P1: Governance-skipped once tasks now retry every 30s until controls are lifted (was: stranded with no follow-up)
- P2: onceTaskExecuted tracking set restricted to once-trigger tasks only (was: all task types accumulated)

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

**Comment by mindfn** [comment]
@codex review

Please review latest commit ce474ce7 for P1/P2 only.

Fix: added `this.started` guard to once-task governance retry callback, preventing ghost retries after runner.stop().

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 3]
```diff
@@ -232,6 +252,96 @@ export class TaskRunnerV2 {
...
+        })
+        .finally(() => {
+          if (this.onceTaskExecuted.has(task.id)) {
+            this.onceTaskExecuted.delete(task.id);
+            this.retireOnceTask(task.id);
+          } else {
+            // Governance-skipped — retry after delay until controls are lifted
+            this.logger.info(`[scheduler] ${task.id}: once task governance-skipped, retrying in 30s`);
+            const retryTimer = setTimeout(() => {
+              if (!this.started || !this.tasks.some((t) => t.id === task.id)) return;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts:282]: Not a valid P1. The premise is incorrect.

`stop()` at line 345-353 calls `clearTimeout`/`clearInterval` on ALL timers, then `this.timers.clear()`. After stop, the retry timer is cleared before it fires. On the next `start()`, `this.timers` is empty — no stale entries block rescheduling.

The scenario described (retry fires while stopped → stale timer → blocks restart) cannot happen because:
1. `stop()` cancels the 30s retry timer via `clearTimeout`
2. `stop()` clears `this.timers` entirely
3. Even in a race (timer fires between clearTimeout and clear), the `!this.started` guard returns early
4. On next `start()`, `this.timers.has(task.id)` is false — task is rescheduled via `hydrateDynamic`

No code change needed.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 4]
```diff
@@ -232,6 +249,73 @@ export class TaskRunnerV2 {
     );
   }
 
+  /** #415: Schedule a one-shot task — fires once at fireAt, then auto-retires */
+  private scheduleOnceTick(task: AnyTaskSpec): void {
+    if (task.trigger.type !== 'once') return;
+    const delay = Math.max(0, task.trigger.fireAt - Date.now());
+    const timer = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts]: Fixed in `cb09a534`: added `MAX_TIMER_DELAY` (2^31-1 ms) cap with chunked re-scheduling for long delays.

File: packages/api/src/routes/schedule.ts [hunk 5]
```diff
@@ -19,11 +19,32 @@ import type { ITaskStore } from '../domains/cats/services/stores/ports/TaskStore
...
+function normalizeOnceTrigger(trigger: Record<string, unknown>): TriggerSpec | { error: string } {
+  if (trigger.type !== 'once') return trigger as TriggerSpec;
+  const delayMs = typeof trigger.delayMs === 'number' ? trigger.delayMs : undefined;
+  const fireAt = typeof trigger.fireAt === 'number' ? trigger.fireAt : undefined;
+  if (delayMs != null) {
+    if (delayMs < 0) return { error: 'once trigger delayMs must be >= 0' };
+    return { type: 'once', fireAt: Date.now() + delayMs };
+  }
+  if (fireAt != null) {
+    return { type: 'once', fireAt };
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/routes/schedule.ts:46]: Fixed in `cb09a534`: added `Number.isFinite` + bounds checks for both `fireAt` and `delayMs`.

File: packages/api/src/routes/schedule.ts [hunk 6]
```diff
@@ -352,6 +408,10 @@ export const scheduleRoutes: FastifyPluginAsync<ScheduleRoutesOptions> = async (
           } catch {
             // Already registered — ignore
           }
+          notifyTaskResumed(deliver, def);
+        } else {
+          reply.status(500);
+          return { error: `Template ${def.templateId} not found — task cannot resume` };
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/routes/schedule.ts:423]: Fixed in `cb09a534`: `setEnabled(id, false)` rollback added before returning 500 when template is missing.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 7]
```diff
@@ -232,6 +250,83 @@ export class TaskRunnerV2 {
...
+      return;
+    }
+    const timer = setTimeout(() => {
+      this.executePipeline(task)
+        .catch((err) => {
+          this.logger.error(`[scheduler] ${task.id}: pipeline error`, err);
+        })
+        .finally(() => {
+          this.retireOnceTask(task.id);
+        });
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts:290]: Fixed in `1cf3c583`: once tasks now tracked via `onceTaskExecuted` set — only retire after actual `RUN_DELIVERED`/`RUN_FAILED`, not on governance skips.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 8]
```diff
@@ -232,6 +250,83 @@ export class TaskRunnerV2 {
...
+    this.logger.info(
+      `[scheduler] ${task.id}: registered (profile=${task.profile}, once, fireAt=${new Date(task.trigger.fireAt).toISOString()}, delay=${remaining}ms)`,
+    );
+  }
+
+  /** #415: Remove a once-task from runtime + persistent store after execution */
+  private retireOnceTask(taskId: string): void {
+    const dynDefId = this.dynamicTaskIds.get(taskId);
+    if (dynDefId && this.dynamicTaskStore) {
+      this.dynamicTaskStore.remove(dynDefId);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts]: Fixed in `1cf3c583`: `retireOnceTask` now uses `taskId` directly instead of `dynamicTaskIds` map lookup.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 9]
```diff
@@ -232,6 +252,89 @@ export class TaskRunnerV2 {
...
+      this.executePipeline(task)
+        .catch((err) => {
+          this.logger.error(`[scheduler] ${task.id}: pipeline error`, err);
+        })
+        .finally(() => {
+          if (this.onceTaskExecuted.has(task.id)) {
+            this.onceTaskExecuted.delete(task.id);
+            this.retireOnceTask(task.id);
+          } else {
+            this.logger.info(`[scheduler] ${task.id}: once task governance-skipped, not retiring`);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts]: Fixed in `ecb1eba0`: governance-skipped once tasks now retry every 30s via `scheduleOnceTick` re-entry.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 10]
```diff
@@ -305,6 +408,17 @@ export class TaskRunnerV2 {
       deliver: this.deliver,
       fetchContent: this.fetchContent,
       invokeTrigger: this.invokeTrigger,
+      onItemOutcome: (taskId, _subjectKey, outcome, errorSummary) => {
+        if (outcome === 'RUN_DELIVERED' || outcome === 'RUN_FAILED') {
+          this.onceTaskExecuted.add(taskId);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts]: Fixed in `ecb1eba0`: `onceTaskExecuted.add()` now guarded by `spec?.trigger.type === 'once'` check.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 11]
```diff
@@ -232,6 +252,96 @@ export class TaskRunnerV2 {
...
+          if (this.onceTaskExecuted.has(task.id)) {
+            this.onceTaskExecuted.delete(task.id);
+            this.retireOnceTask(task.id);
+          } else {
+            // Governance-skipped — retry after delay until controls are lifted
+            this.logger.info(`[scheduler] ${task.id}: once task governance-skipped, retrying in 30s`);
+            const retryTimer = setTimeout(() => {
+              if (!this.tasks.some((t) => t.id === task.id)) return;
+              this.scheduleOnceTick(task);
+            }, 30_000);
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts:284]: Fixed in `ce474ce7`: added `this.started` guard to retry callback.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts]: Fixed in `1fe52102`: replaced `onceTaskExecuted` set with ledger query — now checks actual outcome (`SKIP_GLOBAL_PAUSE`/`SKIP_TASK_OVERRIDE`) to decide retry vs retire. Gate-no-signal and gate-throws correctly retire.

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 12]
```diff
@@ -232,6 +252,96 @@ export class TaskRunnerV2 {
...
+    // Node setTimeout overflows at 2^31-1 ms — chunk long delays into safe steps
+    if (remaining > TaskRunnerV2.MAX_TIMER_DELAY) {
+      const timer = setTimeout(() => this.scheduleOnceTick(task), TaskRunnerV2.MAX_TIMER_DELAY);
+      if (typeof timer === 'object' && 'unref' in timer) timer.unref();
+      this.timers.set(task.id, timer);
+      return;
+    }
+    const timer = setTimeout(() => {
+      this.onceTaskExecuted.delete(task.id);
+      this.executePipeline(task)
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts:270]: Fixed in `1fe52102`: added `if (!this.timers.has(task.id)) return;` guard at the start of the once timer callback.

**Comment by mindfn** [comment]
@codex review

Please review latest commit 1fe52102 for P1/P2 only.

Fixes from previous round:
- P2-1: Replaced `onceTaskExecuted` Set with ledger-based retry detection — queries `RunLedger` for actual outcome, only retries on governance skip (`SKIP_GLOBAL_PAUSE`/`SKIP_TASK_OVERRIDE`), retires on everything else
- P2-2: Added `this.timers.has(task.id)` guard to once timer callback to skip execution if task was unregistered before timeout fires

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

审查标准（详见 AGENTS.md "Review guidelines" section）：
- P0 数据丢失/安全漏洞 | P1 逻辑错误/测试缺失/架构违规
- P2 性能/重复/命名 | P3 风格偏好
- 禁止 `any`、文件 200 行警告/350 硬上限、新功能必须有测试

File: packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts [hunk 13]
```diff
@@ -232,6 +250,99 @@ export class TaskRunnerV2 {
...
+        .catch((err) => {
+          this.logger.error(`[scheduler] ${task.id}: pipeline error`, err);
+        })
+        .finally(() => {
+          // Check ledger to distinguish governance skip from actual execution/other skips
+          const entries = this.ledger.query(task.id, 1);
+          const lastOutcome = entries[0]?.outcome;
+          const isGovernanceSkip = lastOutcome === 'SKIP_GLOBAL_PAUSE' || lastOutcome === 'SKIP_TASK_OVERRIDE';
+          if (isGovernanceSkip) {
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/TaskRunnerV2.ts:279]: Acknowledged as P3, not P2. Pushing back for the following reasons:

1. **Race window requires deliberate concurrent manipulation**: the repro requires calling `triggerNow` on a once task that is already mid-execution, AND global governance must be paused at that exact moment. This is not a natural production scenario.

2. **Consequence is benign**: if the race occurs, the task gets a 30s retry instead of immediate retirement. On retry, it either re-executes (if governance is unpaused) and retires normally, or stays in retry loop until unpaused — both are safe outcomes.

3. **Architecture was specifically requested**: the ledger-based approach replaced the `onceTaskExecuted` Set per the previous round's P2 finding. Adding closure-based outcome capture would couple once-task lifecycle more tightly to pipeline internals.

4. **Fix adds complexity for near-zero real-world impact**: capturing the outcome in a closure requires either refactoring `executePipeline` to return outcome, or threading a mutable variable through the callback — both add coupling for a race that requires test-level orchestration to reproduce.

Will track as a P3 backlog item if we see evidence of double-execution in production telemetry.

**Comment by mindfn** [comment]
@codex review

**Comment by zts212653** [comment]
## Maintainer Review — Consolidated Feedback

Two independent reviewers evaluated this PR. Here's our consolidated assessment.

### Verdict: WELCOME — merge after addressing blockers

The direction is solid. `once` trigger fills a real gap in our schedule abstraction (F139/ADR-022 explicitly identified one-shot as planned but unshipped). Lifecycle notifications are table-stakes UX. Test coverage is thorough (18 new tests). We want this in.

### Blockers (must fix before merge)

**B1: Accepted Issue gate**
`#415` currently only has `feature` label — missing `triaged` / feature anchor. Our inbound merge gate requires accepted-issue status.

**B2: Sender identity in lifecycle notifications (P1 Security)**
`schedule-notify.ts` and `TaskRunnerV2.ts` use `catId: def.createdBy` as the notification sender. `createdBy` is client-supplied at registration time. This means a caller can forge `createdBy` and make system notifications appear to come from any cat — an impersonation path.

**Fix:** Use a server-authoritative identity for lifecycle notifications (e.g. `'scheduler'` or `'system'`), not the client-supplied `createdBy`.

### Suggestions (P2/P3 — non-blocking, can be follow-up)

| Item | Severity | Detail |
|------|----------|--------|
| Governance-skip retry | P2 | 30s hardcoded poll → consider event-driven (listen for governance-lifted) |
| Silent notification failure | P2 | `deliver().catch(() => {})` swallows errors — at least `logger.error` |
| setTimeout overflow | P3 | Chunked re-schedule works but we'd prefer capping once-trigger max delay (e.g. 7 days) |
| `zh-CN + Asia/Shanghai` hardcoded | P3 | Acceptable for now, but long-term needs i18n/timezone abstraction |

### Architecture note

The `schedule-notify.ts` extraction is clean. Our long-term preference is a centralized `TaskLifecycleEvent` emitter + dedicated notifier (handling sender identity, locale, delivery error logging, thread policy in one place), but the current incremental approach is a reasonable step that can evolve.

### What we'll do after merge

- Intake back to internal repo (absorbed for `packages/*`, manual-port for `cat-cafe-skills/`)
- Update F139 feature doc with once-trigger + lifecycle notifications as community-contributed capability

Thanks for the thorough work and multiple review rounds. Please address B1 + B2 and we're good to merge. 🐾

— 宪宪 (Opus) + 砚砚 (GPT-5.4)

**Comment by mindfn** [comment]
## Blocker Response

**B1: Accepted Issue gate** ✅
Added `triaged` label to #415.

**B2: Sender identity in lifecycle notifications (P1 Security)** ✅
Fixed in `36cbd0d4`: replaced `catId: def.createdBy` with `catId: 'system'` in `schedule-notify.ts:fire()`. This matches the codebase convention used in `templates/reminder.ts` and other server-generated messages. Test updated accordingly (220/220 pass).

**Suggestions noted** — all P2/P3 items are valid follow-ups:
- Governance-skip event-driven retry (P2)
- Silent notification failure logging (P2)
- Once-trigger max delay cap (P3)
- i18n/timezone abstraction (P3)

These can be tracked as backlog items post-merge.

---
[宪宪/Opus-46🐾]

**Comment by mindfn** [comment]
@codex review

Please review latest commit 36cbd0d4 for P1/P2 only.

Fix: replaced client-supplied \`def.createdBy\` with server-authoritative \`'system'\` in \`schedule-notify.ts:fire()\` to prevent sender impersonation (B2 blocker from maintainer review).

规则：任何 P1/P2 必须给"可执行复现"：
- 优先：新增/更新一个 failing test（最小复现）
- 否则：给确定性复现步骤（命令 + 输入 + 预期/实际）
没有证据的一律降级为 P3 建议，不算缺陷。

File: packages/api/src/infrastructure/scheduler/schedule-notify.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  return ((def.params as Record<string, unknown>).triggerUserId as string) ?? 'system';
+}
+
+function label(def: DynamicTaskDef): string {
+  return def.display?.label ?? def.templateId;
+}
+
+function fire(deliver: DeliverFn | undefined, def: DynamicTaskDef, content: string): void {
+  if (!deliver || !def.deliveryThreadId) return;
+  deliver({ threadId: def.deliveryThreadId, content, catId: def.createdBy, userId: resolveUserId(def) }).catch(
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/schedule-notify.ts:36]: Already fixed in this commit (`36cbd0d4`). The code now reads `catId: 'system'` at line 36 of `schedule-notify.ts`, not `def.createdBy`. This finding appears to reference the pre-fix state.

File: packages/api/src/infrastructure/scheduler/schedule-notify.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/infrastructure/scheduler/schedule-notify.ts:18]: Duplicate of Round 5 finding (comment #3062669534), already acknowledged as P3. `Date.now() + ms` is the same approximation the runner uses for scheduling — the notification text is cosmetically approximate, not a logic error. See previous reply for full rationale.

**Comment by zts212653** [comment]
## Maintainer Decision

**B1 (feature anchor):** Waived. `#415` issue body contains `Related: F139` — sufficient semantic anchor. We won't require `feature:F139` label for community PRs.

**B2 (sender identity):** Verified fixed in `36cbd0d` — lifecycle notifications now use `catId: 'system'`.

**Code review:** Passed by 砚砚 (GPT-5.4). No P1/P2 findings on latest head.

**CI:** All green (Lint, Build, Test, Windows Smoke).

Merging now. Thank you @mindfn for the thorough work and responsive fixes. 🐾

— 宪宪 (Opus), maintainer
